README.md
[![Travis-CI Build Status](https://travis-ci.org/magosil86/getmstatistic.svg?branch=master)](https://travis-ci.org/magosil86/getmstatistic)
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/magosil86/getmstatistic/master/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/magosil86/getmstatistic.svg)](https://github.com/magosil86/getmstatistic/issues)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/getmstatistic)](https://cran.r-project.org/package=getmstatistic)
[![Coverage status](https://codecov.io/gh/magosil86/getmstatistic/branch/master/graph/badge.svg)](https://codecov.io/github/magosil86/getmstatistic?branch=master)
[![CRAN_Logs_Rstudio](https://cranlogs.r-pkg.org/badges/grand-total/getmstatistic)](http://cran.rstudio.com/web/packages/getmstatistic/index.html)

<!--- [![CRAN_Logs_Rstudio](http://cranlogs.r-pkg.org/badges/getmstatistic)](http://cran.rstudio.com/web/packages/getmstatistic/index.html) --->


# [getmstatistic]() <img src="getmstatistic-striking-image.png" align="right" />

_M_ - An aggregate statistic, to identify systematic heterogeneity patterns and their direction of effect in meta-analysis 
 
## Overview
 
 _M_ quantitatively describes systematic (non-random) heterogeneity patterns acting across multiple variants in a GWAS meta-analysis. It's primary use is to identify outlier studies, which either show "null" effects or consistently show stronger or weaker genetic effects than average across, the panel of variants examined in a meta-analysis.

 _M_ differs from conventional heterogeneity metrics (Q-statistic, I<sup>2</sup>), in that, it measures heterogeneity across multiple independently associated variants, whilst (Q-statistic, I<sup>2</sup>) measure heterogeneity at individual variants. Essentially, _M_ measures systematic heterogeneity, whilst (Q-statistic, I<sup>2</sup>) measure random variant-specific heterogeneity.

### Sources of systematic heterogeneity
Systematic heterogeneity can arise in a meta-analysis due to differences in the study characteristics of participating studies. Some of the differences may include: ancestry, allele frequencies, phenotype definition, age-of-disease onset, family-history, gender, linkage disequilibrium and quality control thresholds.

### Practical benefits of exploring systematic heterogeneity

* Reveal studies showing systematically weaker effects than average which could lower the power of a meta-analysis to detect genetic signals. For example, outlier studies that pass typical quality control checks (genotype call rate, Hardy-Weinberg equilibrium cutoffs, genomic control) but might show no association with phenotype of interest due to faulty genotype data (e.g. flipped alleles and/or strands, incorrect minor allele frequencies).

* Reveal studies showing systematically stronger effects than average which can elucidate biologically important differences among the studies e.g. sexual dimorphism or sub-phenotype variability.



## Installation: Stata

* Install getmstatistic [**Stata** command](https://github.com/magosil86/getmstatistic) from SSC

1. Start stata
2. To install the getmstatistic stata module: `ssc install getmstatistic`
3. To get the example dataset in your current working directory: `net get getmstatistic`
4. You're all set, getmstatistic is installed, try some of the examples in the getmstatistic help file

Tip! getmstatistic depends on the following user-written Stata commands which can be installed
 using `ssc install package-name` or findit `package-name`:

* latabstat
* metareg
* savesome
* tabstat
* qqvalue

A full list of getmstatistic dependencies can be found in the help file.

---

* Install getmstatistic [**Stata** command](https://github.com/magosil86/getmstatistic) manually

* Download the getmstatistic zip file: [getmstatistic_0.1.1_stata_ssc.zip](https://github.com/magosil86/getmstatistic/blob/master/getmstatistic_0.1.1_stata/getmstatistic_0.1.1_stata_ssc.zip)

```
1. Unzip the folder

Tip! the folder should contain the following files: getmstatistic.ado getmstatistic.sthlp heartgenes214.dta

2. Start stata

3. Locate your personal directory where stata stores user generated files by typing: `sysdir`
sysdir

Tip! on mac the ado/personal directory is likely to be at: ~/Library/Application Support/Stata/ado/personal/
for linux: ~/ado/personal/ 
for windows: c:\ado\personal\

4. Copy getmstatistic.ado and getmstatistic.sthlp to the g sub-directory in personal

Tip! if the g sub-directory does not exist, that just means you do not have user generated commands
that start with the letter g. In that case create a folder named g in the personal directory.

5. Type help getmstatistic in Stata to open the getmstatistic help file.

6. Load the example dataset: heartgenes214.dta

7. You're all set, getmstatistic is installed, try some of the examples in the getmstatistic help file

```

## Installation: getmstatistic [**R** package](https://github.com/magosil86/getmstatistic)

```{r}
# To install the release version from CRAN:
install.packages("getmstatistic")

# Load libraries
library(getmstatistic)  # for calculating M statistics
library(gridExtra)      # for generating tables


# To install the development version from GitHub:

# install devtools
install.packages("devtools")

# install getmstatistic
library(devtools)
devtools::install_github("magosil86/getmstatistic")

# Load libraries
library(getmstatistic)  # for calculating M statistics
library(gridExtra)      # for generating tables

```


## Usage

*  Take a look at an [example workflow](https://github.com/magosil86/getmstatistic/blob/master/vignettes/getmstatistic-tutorial.md)

## Details

* Essentially, _M_ statistics are computed by aggregating standardized predicted random effects (SPREs). To read up about the statistical theory behind the _M_ statistic see:

Magosi LE, Goel A, Hopewell JC, Farrall M, on behalf of the CARDIoGRAMplusC4D Consortium (2017) Identifying systematic heterogeneity patterns in genetic association meta-analysis studies. PLoS Genet 13(5): e1006755. [https://doi.org/10.1371/journal.pgen.1006755](https://doi.org/10.1371/journal.pgen.1006755).


## Getting help

To suggest new features, learn about getmstatistic updates, report bugs, ask questions about the mstatistic, or just interact with other users, sign up to the [getmstatistic](https://groups.google.com/forum/#!forum/getmstatistic) mailing list.


## Code of conduct
Contributions are welcome. Please observe the [Contributor Code of Conduct](https://github.com/magosil86/getmstatistic/blob/master/CONDUCT.md) when participating in this project.

## Citation
Magosi LE, Goel A, Hopewell JC, Farrall M, on behalf of the CARDIoGRAMplusC4D Consortium (2017) Identifying systematic heterogeneity patterns in genetic association meta-analysis studies. PLoS Genet 13(5): e1006755. [https://doi.org/10.1371/journal.pgen.1006755](https://doi.org/10.1371/journal.pgen.1006755).


## Acknowledgements.
Roger M. Harbord’s metareg command for computation of standardized predicted random effects which are then incorporated into calculations for the _M_ statistics. Harbord, R. M., & Higgins, J. P. T. (2008). Meta-regression in Stata. Stata Journal 8: 493‚Äì519.


## Authors.
Lerato E. Magosi, Jemma C. Hopewell and Martin Farrall.

## Maintainer.
Lerato E. Magosi lmagosi@well.ox.ac.uk or magosil86@gmail.com

## License

See the [LICENSE](https://github.com/magosil86/getmstatistic/blob/master/LICENSE) file.



NEWS.md
# getmstatistic 0.2.2

## Minor improvements

* Added rmarkdown to suggested dependencies


# getmstatistic 0.2.1

## Minor improvements

* Exported the draw_table function to allow 
  running of \donttest examples.
  
* Generating plots has been made optional to provide
  more flexibility
  
* Plots generated when testing examples are saved in a 
  temporary directory


# getmstatistic 0.2.0

## Minor improvements

* Updated maintainer email address


# getmstatistic 0.1.1

## Minor improvements

* Separated the further arguments and example sections in 
  the draw_table function for neater rendering of the pdf manual
  
* Improved wrapping of titles for the M statistic tables

* Updated citation

R/data.R
#' heartgenes214.
#'
#' heartgenes214 is a multi-ethnic GWAS meta-analysis dataset for coronary
#' artery disease.
#'
#' It comprises summary data (effect-sizes and their corresponding standard 
#'     errors) for 48 studies (68,801 cases and 123,504 controls), at 214 lead 
#'     variants independently associated with coronary artery disease 
#'     (P < 0.00005, FDR < 5\%). Of the 214 lead variants, 44 are genome-wide 
#'     significant (p < 5e-08). The meta-analysis dataset is based on 
#'     individuals of: African American, Hispanic American, East Asian, 
#'     South Asian, Middle Eastern and European ancestry. 
#'
#' The study effect-sizes have been flipped to ensure alignment of the 
#'     effect alleles.
#'
#' Standard errors were genomically corrected at the study-level.
#'
#'
#' @format A data frame with seven variables:
#' \describe{
#' \item{\code{beta_flipped}}{Effect-sizes expressed as log odds ratios. Numeric}
#' \item{\code{gcse}}{Standard errors}
#' \item{\code{studies}}{Names of participating studies}
#' \item{\code{variants}}{Names of genetic variants/SNPs}
#' \item{\code{cases}}{Number of cases in each participating study}
#' \item{\code{controls}}{Number of controls in each participating study}
#' \item{\code{fdr214_gwas46}}{Flag indicating GWAS significant variants, 
#'       1: Not GWAS-significant, 2: GWAS-significant}
#' }
#'
#' @source Magosi LE, Goel A, Hopewell JC, Farrall M, on behalf of the 
#'         CARDIoGRAMplusC4D Consortium (2017) Identifying systematic 
#'         heterogeneity patterns in genetic association meta-analysis studies. 
#'         PLoS Genet 13(5): e1006755. https://doi.org/10.1371/journal.pgen.1006755.
#'
#'         \url{https://magosil86.github.io/getmstatistic/}
#'
"heartgenes214"

R/util_draw_table.R
#' Helper function to draw table grobs. 
#' 
#' \code{draw_table()} Pre and post version: 2.0.0 gridExtra packages
#' handle drawing tables differently. \code{draw_table()} determines
#' the installed version of gridExtra and applies the appropriate
#' syntax. If gridExtra version < 2.0.0 then it uses old gridExtra 
#' syntax to build table Grob(graphical object) else uses new syntax. 
#' \code{draw_table()}
#' 
#' prints tables without rownames.
#'
#' @section Acknowledgements: Thanks to Ryan Welch, 
#' https://github.com/welchr/LocusZoom/issues/16
#'
#'
#' @param body A dataframe. Table body.
#' @param heading A string. Table title.
#' @param ... Further arguments to control the gtable.
#'
#' @examples
#' 
#' library(gridExtra)
#'
#' \dontrun{
#'
#' # Table of iris values
#' iris_dframe <- head(iris)
#' title_iris_dframe <- paste("Table: Length and width measurements (cm) of sepals and petals,",
#'                             "for 50 flowers from 3 species of iris (setosa, versicolor,", 
#'                             "and virginica).\n", sep = " ")
#' # Wrap title text at column 60
#' title_iris_dframe <- sapply(strwrap(title_iris_dframe, width = 60, simplify = FALSE), 
#'                             paste, collapse = "\n")
#' # Draw table
#' table_influential_studies <- draw_table(body = iris_dframe, heading = title_iris_dframe)
#' 
#' # Table of mtcars values
#' mtcars_dframe <- head(mtcars)
#' title_mtcars_dframe <- paste("Table: Motor Trend US magazine (1974) automobile statistics", 
#'                              "for fuel consumption, \nautomobile design and performance.\n", 
#'                              sep = " ")
#' # Wrap title text at column 60
#' title_mtcars_dframe <- sapply(strwrap(title_mtcars_dframe, width = 60, simplify = FALSE), 
#'                               paste, collapse = "\n")
#' # Draw table
#' table_influential_studies <- draw_table(body = mtcars_dframe, heading = title_mtcars_dframe)
#' }

#' @export
draw_table <- function(body, heading, ...) {

  ge_version <- as.character(utils::packageVersion("gridExtra"))
  vcomp = utils::compareVersion(ge_version,"2.0.0")
  
  if (vcomp == -1) {
    table_old <- gridExtra::tableGrob(
      
      body,
      show.rownames = FALSE,
      gpar.corefill = grid::gpar(fill = "white", col = "black"),
      gpar.colfill = grid::gpar(fill = "white", col = "black"),
      ...
    )
    
    # Assigning the var h the height of our table
	height_table_old <- grid::grobHeight(table_old)

	title_old <- grid::textGrob(heading, y=grid::unit(0.5, "npc") + 0.7*height_table_old,vjust=0,gp=grid::gpar(fontsize=12.0))
	gt_table_old <- grid::gTree(children=grid::gList(table_old, title_old))
	
	print(grid::grid.newpage())
	print(grid::grid.draw(gt_table_old))


  } else {
  
    tt <- gridExtra::ttheme_default(
        core=list(bg_params = list(fill = "white", col="black")),
        colhead=list(bg_params = list(fill = "white", col="black")))
  
    table <- gridExtra::tableGrob(
      
      body,
      theme = tt,
      rows = NULL,
      ...
    )
    
    title <- grid::textGrob(heading, gp=grid::gpar(fontsize=12), vjust=-1)
    
    padding <- grid::unit(0.2,"npc")
    
    table <- gtable::gtable_add_rows(table, heights = grid::grobHeight(title) + padding, pos = 0)
    
    table <- gtable::gtable_add_grob(table, title, t=1, b=nrow(table), l=1, r=ncol(table))
   
   #rectangle grob dimensions: height:row3 --> nrows, width:col1 --> ncols
    table <- gtable::gtable_add_grob(table, grobs = grid::rectGrob(gp = grid::gpar(fill = NA, lwd = 2)), t = 3, b = nrow(table), l = 1, r = ncol(table))

   #rectangle grob dimensions: height:row2 --> row2, width:col1 --> ncols
    table <- gtable::gtable_add_grob(table, grobs = grid::rectGrob(gp = grid::gpar(fill = NA, lwd = 2)), t = 2, b = 2, l = 1, r = ncol(table))
 
    print(grid::grid.newpage())
    print(grid::grid.draw(table))


  }
}



R/driver_compute_mstatistics.R
# Goal: Generate and export S3 class/methods for getmstatistic and document using roxygen tags
#
#' Quantifying Systematic Heterogeneity in Meta-Analysis.
#'
#' \code{getmstatistic} computes \emph{M} statistics to assess the contribution
#' of each participating study in a meta-analysis. The \emph{M} statistic 
#' aggregates heterogeneity information across multiple variants to, identify 
#' systematic heterogeneity patterns and their direction of effect in 
#' meta-analysis. It's primary use is to identify outlier studies, which either 
#' show "null" effects or consistently show stronger or weaker genetic effects 
#' than average, across the panel of variants examined in a GWAS meta-analysis. 
#'
#' In contrast to conventional heterogeneity metrics (Q-statistic, I-squared and 
#' tau-squared) which measure random heterogeneity at individual variants, 
#' \emph{M} measures systematic (non-random) heterogeneity across multiple 
#' independently associated variants. 
#'
#' Systematic heterogeneity can arise in a meta-analysis due to differences in 
#' the study characteristics of participating studies. Some of the differences 
#' may include: ancestry, allele frequencies, phenotype definition, 
#' age-of-disease onset, family-history, gender, linkage disequilibrium and 
#' quality control thresholds. See the getmstatistic website for statistical 
#' theory, documentation and examples.
#'
#' \code{getmstatistic} uses summary data i.e. study effect-sizes and their 
#' corresponding standard errors to calculate \emph{M} statistics (One \emph{M} 
#' for each study in the meta-analysis).
#'
#' In particular, \code{getmstatistic} employs the inverse-variance weighted 
#' random effects regression model provided in the \code{metafor} R package 
#' to extract SPREs (standardized predicted random effects) which are then 
#' aggregated to formulate \emph{M} statistics.
#' 
#' 
#' @seealso \code{\link[metafor]{rma.uni}} function in \code{metafor} for random
#'     effects model, and \url{https://magosil86.github.io/getmstatistic/} for 
#'     getmstatistic website. 
#'
#' @aliases getm getmstat getmstatistic Rgetmstatistic
#'
#' @param beta_in A numeric vector of study effect-sizes e.g. log odds-ratios. 
#' @param lambda_se_in A numeric vector of standard errors, genomically 
#'        corrected at study-level. 
#' @param study_names_in A character vector of study names. 
#' @param variant_names_in A character vector of variant names e.g. rsIDs.
#' @param tau2_method A character scalar, method to estimate heterogeneity: 
#'        either "DL" or "REML" (Optional). Note: The REML method uses the 
#'        iterative Fisher scoring algorithm (step length = 0.5, maximum 
#'        iterations = 10000) to estimate tau2.
#' @param x_axis_increment_in A numeric scalar, value by which x-axis 
#'        of M scatterplot will be incremented (Optional).
#' @param x_axis_round_in A numeric scalar, value to which x-axis labels of 
#'        M scatterplot will be rounded (Optional).
#' @param produce_plots A boolean to generate plots (optional).
#' @param save_dir A character scalar specifying a path to the directory where
#'        plots should be stored (optional). Required if produce_plots = TRUE.
#' @param verbose_output An optional boolean to display intermediate output.
#' @param ... Further arguments.
#'
#' @return Returns a list containing:
#' \itemize{
#'   \item Mstatistic_expected_mean   , A numeric scalar for the expected mean for M
#'   \item Mstatistic_expected_sd     , A numeric scalar for the expected standard deviation 
#'         for M
#'   \item number_studies             , A numeric scalar for the number of studies
#'   \item number_variants            , A numeric scalar for the number of variants
#'   \item Mstatistic_crit_alpha_0_05 , A numeric scalar of the critical M value at the 5
#'         percent significance level.
#'   \item M_dataset (dataframe) A dataset of the computed M statistics, which
#'         includes the following fields:
#'   \itemize{
#'         \item M           , Mstatistic
#'         \item M_sd        , standard deviation of M
#'         \item M_se        , standard error of M
#'         \item lowerbound  , lowerbound of M 95% confidence interval
#'         \item upperbound  , upperbound of M 95% confidence interval
#'         \item bonfpvalue  , 2-sided bonferroni pvalues of M
#'         \item qvalue      , false discovery rate adjusted pvalues of M
#'         \item tau2        , tau_squared, DL estimates of between-study 
#' heterogeneity
#'         \item I2          , I_squared, proportion of total variation due to 
#' between study variance
#'         \item Q           , Cochran's Q
#'         \item xb          , fitted values excluding random effects
#'         \item usta        , standardized predicted random effect (SPRE)
#'         \item xbu         , fitted values including random effects
#'         \item stdxbu      , standard error of prediction (fitted values) 
#' including random effects
#'         \item hat         , diagonal elements of the projection hat matrix
#'         \item study       , study numbers    
#'         \item snp         , variant numbers
#'         \item beta_mean   , average variant effect size
#'         \item oddsratio   , average variant effect size as oddsratio
#'         \item beta_n   , number of variants in each study
#'         }
#'   \item influential_studies_0_05 (dataframe) A dataset of influential studies 
#'        significant at the 5 percent level.
#'   \item weaker_studies_0_05 (dataframe) A dataset of under-performing studies 
#'        significant at the 5 percent level.
#' }
#'
#' @examples
#' library(getmstatistic)
#' library(gridExtra)
#'
#' 
#' # Basic M analysis using the heartgenes214 dataset.
#' # heartgenes214 is a multi-ethnic GWAS meta-analysis dataset for coronary artery disease.
#' # To learn more about the heartgenes214 dataset ?heartgenes214
#'
#' # Running an M analysis on 20 GWAS significant variants (p < 5e-08) in the first 10 studies
#'
#' heartgenes44_10studies <- subset(heartgenes214, studies <= 10 & fdr214_gwas46 == 2) 
#' heartgenes20_10studies <- subset(heartgenes44_10studies, 
#'     variants %in% unique(heartgenes44_10studies$variants)[1:20])
#' 
#' # Set directory to store plots, this can be a temporary directory
#' # or a path to a directory of choice e.g. plots_dir <- "~/Downloads"
#' plots_dir <- tempdir()
#'
#' getmstatistic_results <- getmstatistic(heartgenes20_10studies$beta_flipped, 
#'                                         heartgenes20_10studies$gcse, 
#'                                         heartgenes20_10studies$variants, 
#'                                         heartgenes20_10studies$studies,
#'                                         save_dir = plots_dir)
#' getmstatistic_results
#'
#' # Explore results generated by getmstatistic function
#'
#' # Retrieve dataset of M statistics
#' dframe <- getmstatistic_results$M_dataset
#' 
#' \donttest{
#' 
#' str(dframe)
#'}
#'
#' # Retrieve dataset of stronger than average studies (significant at 5% level)
#' getmstatistic_results$influential_studies_0_05
#' 
#' # Retrieve dataset of weaker than average studies (significant at 5% level)
#' getmstatistic_results$weaker_studies_0_05
#'
#' # Retrieve number of studies and variants
#' getmstatistic_results$number_studies
#' getmstatistic_results$number_variants
#' 
#' # Retrieve expected mean, sd and critical M value at 5% significance level
#' getmstatistic_results$M_expected_mean
#' getmstatistic_results$M_expected_sd
#' getmstatistic_results$M_crit_alpha_0_05
#'
#' # To view plots stored in a temporary directory, call `tempdir()` to view the directory path 
#' tempdir()
#'
#'
#' # Additional examples: These take a little bit longer to run
#' 
#' \dontrun{
#'
#' # Set directory to store plots, this can be a temporary directory
#' # or a path to a directory of choice e.g. plots_dir <- "~/Downloads"
#' plots_dir <- tempdir()
#'
#' # Run M analysis on all 214 lead variants
#' # heartgenes214 is a multi-ethnic GWAS meta-analysis dataset for coronary artery disease.
#' getmstatistic_results <- getmstatistic(heartgenes214$beta_flipped, 
#'                                         heartgenes214$gcse, 
#'                                         heartgenes214$variants, 
#'                                         heartgenes214$studies,
#'                                         save_dir = plots_dir)
#' getmstatistic_results
#'
#'
#' # Subset the GWAS significant variants (p < 5e-08) in heartgenes214
#' heartgenes44 <- subset(heartgenes214, heartgenes214$fdr214_gwas46 == 2)
#'
#' # Exploring getmstatistic options:
#' #     Estimate heterogeneity using "REML", default is "DL"
#' #     Modify x-axis of M scatterplot
#' #     Run M analysis verbosely
#' getmstatistic_results <- getmstatistic(heartgenes44$beta_flipped, 
#'                                         heartgenes44$gcse, 
#'                                         heartgenes44$variants, 
#'                                         heartgenes44$studies,
#'                                         save_dir = plots_dir,
#'                                         tau2_method = "REML",
#'                                         x_axis_increment_in = 0.03, 
#'                                         x_axis_round_in = 3,
#'                                         produce_plots = TRUE,
#'                                         verbose_output = TRUE)
#' getmstatistic_results
#'
#'
#' }
#'
#' @export
getmstatistic <- function(beta_in, lambda_se_in, study_names_in, variant_names_in, ...) UseMethod("getmstatistic")

#' @describeIn getmstatistic Computes M statistics
#' @export
getmstatistic.default <- function(beta_in, lambda_se_in, study_names_in, 
                                  variant_names_in, save_dir = getwd(), tau2_method = "DL",
                                  x_axis_increment_in = 0.02, x_axis_round_in = 2, 
                                  produce_plots = TRUE, verbose_output = FALSE, ...) {

    # Check whether all required variables are present
    if (missing(beta_in))
        stop("Beta values missing, need to specify study effect-sizes.")
 
    if (missing(lambda_se_in))
        stop("Standard errors missing, need to specify standard errors.")

    if (missing(study_names_in))
        stop("Study names missing, need to specify study names.")

    if (missing(variant_names_in))
        stop("Variant names missing, need to specify variant names.")

    
    # Verify datatypes of required variables
    if (!is.numeric(c(beta_in, lambda_se_in))) {

        stop("beta_in and lambda_se_in should be of type, numeric.")

    }


    if (!is.character(c(variant_names_in, study_names_in))) {

        stop("variant_names_in and study_names_in should be of type, character.")

    }
    
    compute_m_results <- compute_m_statistics(beta_in, lambda_se_in, study_names_in, variant_names_in, 
                                              save_dir, tau2_method, x_axis_increment_in, 
                                              x_axis_round_in, produce_plots, verbose_output)
    
    compute_m_results$call <- match.call()
    
    class(compute_m_results) <- "getmstatistic"
    
    compute_m_results
    
}

#' @export
print.getmstatistic <- function(x, ..., verbose_output = FALSE) {

    cat("Call:\n")
    print(x$call)

    cat("\nM_expected_mean:\n")
    print(x$M_expected_mean)

    cat("\nM_expected_sd:\n")
    print(x$M_expected_sd)

    cat("\nM_crit_alpha_0_05:\n")
    print(x$M_crit_alpha_0_05)
    
    cat("\ninfluential_studies_0_05:\n")
    print(x$influential_studies_0_05)
    
    cat("\nweaker_studies_0_05:\n")
    print(x$weaker_studies_0_05)


}
